### ISSUE TITLE
Please add interface IReadOnlySet and make HashSet, SortedSet implement it

### ISSUE THREAD

**Issue by dsaf** [dotnet/runtime#2293]
# Original
Since  `IReadOnlyList` was added the parity between sets and lists has declined. It would be great to re-establish it.

Using it would be an implementation-agnostic way of saying: "here is this read-only collection where items are unique".

Clearly it's needed by many people:

SQL Server: https://msdn.microsoft.com/en-us/library/gg503096.aspx
Roslyn: https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/InternalUtilities/IReadOnlySet.cs
Some Guy In Comments: http://blogs.msdn.com/b/bclteam/archive/2013/03/06/update-to-immutable-collections.aspx

I found this discussion when working on a real world problem where I wanted to use the key collection of a dictionary for read only set operations. In order to support that case here's the API I propose.

# Edit
## Proposed API

```diff
 namespace System.Collections.Generic {
+    public interface IReadOnlySet<out T> : IReadOnlyCollection<T>, IEnumerable, IEnumerable<T> {
+        bool Contains(T value);
+        bool IsProperSubsetOf(IEnumerable<T> other);
+        bool IsProperSupersetOf(IEnumerable<T> other);
+        bool IsSubsetOf(IEnumerable<T> other);
+        bool IsSupersetOf(IEnumerable<T> other);
+        bool Overlaps(IEnumerable<T> other);
+        bool SetEquals(IEnumerable<T> other);
+    }
-    public class HashSet<T> : ICollection<T>, IDeserializationCallback, IEnumerable, IEnumerable<T>, IReadOnlyCollection<T>, ISerializable, ISet<T> {
+    public class HashSet<T> : ICollection<T>, IDeserializationCallback, IEnumerable, IEnumerable<T>, IReadOnlyCollection<T>, ISerializable, ISet<T>, IReadOnlySet<T> {
     }
-    public class SortedSet<T> : ICollection, ICollection<T>, IDeserializationCallback, IEnumerable, IEnumerable<T>, IReadOnlyCollection<T>, ISerializable, ISet<T> {
+    public class SortedSet<T> : ICollection, ICollection<T>, IDeserializationCallback, IEnumerable, IEnumerable<T>, IReadOnlyCollection<T>, ISerializable, ISet<T>, IReadOnlySet<T> {
     }
+    public class ReadOnlySet<T> : ICollection<T>, IEnumerable, IEnumerable<T>, IReadOnlyCollection<T>, IReadOnlySet<T>, ISet<T> {
+        public int Count { get; }
+        public ReadOnlySet(ISet<T> set);
+        public bool Contains(T value);
+        public bool IsProperSubsetOf(IEnumerable<T> other);
+        public bool IsProperSupersetOf(IEnumerable<T> other);
+        public bool IsSubsetOf(IEnumerable<T> other);
+        public bool IsSupersetOf(IEnumerable<T> other);
+        public bool Overlaps(IEnumerable<T> other);
+        public bool SetEquals(IEnumerable<T> other);
+    }
     public class Dictionary<TKey, TValue> {
-        public sealed class KeyCollection : ICollection, ICollection<TKey>, IEnumerable, IEnumerable<TKey>, IReadOnlyCollection<TKey> {
+        public sealed class KeyCollection : ICollection, ICollection<TKey>, IEnumerable, IEnumerable<TKey>, IReadOnlyCollection<TKey>, IReadOnlySet<TKey> {
+            public bool IsProperSubsetOf(IEnumerable<TKey> other);
+            public bool IsProperSupersetOf(IEnumerable<TKey> other);
+            public bool IsSubsetOf(IEnumerable<TKey> other);
+            public bool IsSupersetOf(IEnumerable<TKey> other);
+            public bool Overlaps(IEnumerable<TKey> other);
+            public bool SetEquals(IEnumerable<TKey> other);
         }
     }
 }
```

## Open Questions
* Is adding these methods to `Dictionary<TKey, TValue>.KeyCollection` acceptable given the code size cost for a commonly instantiated generic type? See [here](https://github.com/dotnet/corefx/issues/17275#issuecomment-291636863).
* Should `IReadOnlySet<T>` be implemented in other `Dictionary` `KeyCollection`s such as `SortedDictionary` or `ImmutableDictionary`?

## Updates
* Removed `TryGetValue` as it's not in `ISet<T>` and as such would prevent potentially ever rebasing `ISet<T>` to implement `IReadOnlySet<T>`.
* Added `ReadOnlySet<T>` class which is similar to `ReadOnlyCollection<T>`.

**Comment by MatthewLymer** [comment]
Do it, I dare you!

**Comment by svick** [comment]
@terrajobst 

> We can't make `ISet<T>` extend `IReadOnlySet<T>` (for the same reason that we couldn't do for the other mutable interfaces).

Is this still true even with default interface methods? Does that mean https://github.com/dotnet/corefx/issues/41409 should be closed?

**Comment by terrajobst** [comment]
> @terrajobst
> 
> > We can't make `ISet<T>` extend `IReadOnlySet<T>` (for the same reason that we couldn't do for the other mutable interfaces).
> 
> Is this still true even with default interface methods? Does that mean [dotnet/corefx#41409](https://github.com/dotnet/corefx/issues/41409) should be closed?

We discussed this. We used to think that that DIMs *would* work, but when we walked the solution we concluded that it would result commonly in a shard diamond which would result in an ambiguous match. However, this was recently challenged so I think I have to write it down and make sure it's actually working or not working.

**Comment by Jlalond** [comment]
@terrajobst / @danmosemsft Has Anyone been assigned to this? 

And, @terrajobst to clarify the work we want to achieve is:
```
We should also implement IReadOnlySet<T> on ImmutableSortedSet<T> and ImmutableHashSet<T> (and their builders)
We should scan for other implementations of ISet<T>.
````
As well as implementing the above interfaces on HashSet, SortedSet.

The scanning being just look for anything and bring it up if it looks questionable.

If this is still up for grabs I'd be interested

**Comment by danmosemsft** [comment]
@Jlalond nope, assigned to you.  Thanks for the offer.

**Comment by Jlalond** [comment]
@danmosemsft @terrajobst 
Just an update. I'm working on this, added the interface to private core Lib, just stumbling my way through having collections.generic and immutable pick up on this.

Last question if you know off your head Dan, do I need to make any changes to Mono for this? I'm not insightful on where corefx ends and mono starts. So if you know it could save me from some independent research

**Comment by danmosemsft** [comment]
@Jlalond you should not need to make changes to Mono. Part of the reason for moving the Mono runtime into this repo is to make it seamless to use the same exact libraries with either CoreCLR or the Mono runtime. There is only a small part of the core library that diverges:
coreclr\src\System.Private.CoreLib vs mono\netcore\System.Private.CoreLib. (Most of the core library is shared out of libraries\System.Private.CoreLib). So unless you touch that - you are not affected.

**Comment by Jlalond** [comment]
@danmosemsft Thanks for the clarification, I hope to have this done shortly.

**Comment by Jlalond** [comment]
Hey @danmosemsft just follow up on this. CoreLib is building in the src assemblies I can see the referenced changes. However the ref assemblies seem to not be detecting any changes. This is all that's holding me up, but I can't find any info in the docs. Any people or pointers you can give me so I can get this done (I mean, 5 years later)

**Comment by eiriktsarpalis** [comment]
Addressed by #32488
